src/librustc_typeck/astconv.rs

```diff
@@ -748,6 +743,57 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
                 }
             },
         );
+        if !missing_type_params.is_empty() {
+            let display = missing_type_params.iter()
+                .map(|n| format!("`{}`", n))
+                .collect::<Vec<_>>().join(", ");
+            let mut err = struct_span_err!(tcx.sess, span, E0393,
+                "the type parameter{} {} must be explicitly specified",
+                pluralize!(missing_type_params.len()),
+                display,
+            );
+            err.span_label(self.tcx().def_span(def_id), &format!(
+                "type parameter{} {} must be specified for this",
+                pluralize!(missing_type_params.len()),
+                display,
+            ));
+            let mut suggested = false;
+            if let (Ok(snippet), true) = (
+                tcx.sess.source_map().span_to_snippet(span),
+                // Don't suggest setting the type params if there are some already: the order is
+                // tricky to get right and the user will already know what the syntax is.
+                generic_args.args.is_empty(),
+             ) {
+                if snippet.ends_with('>') {
+                    // The user wrote `Trait<'a, T>` or similar. To provide an accurate suggestion
+                    // we would have to preserve the right order. For now, as clearly the user is
+                    // aware of the syntax, we do nothing.
+                } else {
+                    // The user wrote `Iterator`, so we don't have a type we can suggest, but at
+                    // least we can clue them to the correct syntax `Iterator<Type>`.
+                    err.span_suggestion(
+                        span,
+                        &format!(
+                            "set the type parameter{plural} to the desired type{plural}",
+                            plural=pluralize!(missing_type_params.len()),
+                        ),
+                        format!("{}<{}>", snippet, missing_type_params.join(", ")),
+                        Applicability::HasPlaceholders,
+                    );
+                    suggested = true;
+                }
+            }
+            if !suggested {
+                err.span_label(span, format!(
+                    "missing reference{} to {}",
+                    pluralize!(missing_type_params.len()),
+                    display,
+                ));
+            }
+            err.note(&format!("because of the default `Self` reference, type parameters must be \
+                               specified on object types"));
+            err.emit();
+        }
```

Comment:
Would be good to move this to its own function, especially as this is in a pretty important part of the type system.